include/wil/windowing.h

```diff
@@ -0,0 +1,78 @@
+//*********************************************************
+//
+//    Copyright (c) Microsoft. All rights reserved.
+//    This code is licensed under the MIT License.
+//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+//    ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+//    TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+//    PARTICULAR PURPOSE AND NONINFRINGEMENT.
+//
+//*********************************************************
+#ifndef __WIL_WINDOWING_INCLUDED
+#define __WIL_WINDOWING_INCLUDED
+
+#include <WinUser.h>
+
+#if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP)
+namespace wil
+{
+    namespace details
+    {
+        template<typename TCallback>
+        BOOL CALLBACK LambdaEnumWindowsProc(HWND data, LPARAM lParam) noexcept
+        {
+            const TCallback* pCallback = reinterpret_cast<const TCallback*>(lParam);
+            using result_t = decltype((*pCallback)(data));
+#ifdef __cpp_if_constexpr
+            if constexpr (std::is_void_v<result_t>)
+            {
+                try
+                {
+                    (*pCallback)(data);
+                    return TRUE;
+                }
+                catch (...)
+                {
+                    return FALSE;
+                }
+            }
+            else if constexpr (std::is_same_v<result_t, HRESULT>)

```

Comment:
This makes me feel a little uneasy, but I'm not sure how legitimate that fear is. `HRESULT` is defined as `long` and `BOOL` is defined as `int` (if you're in `minwindef.h`; if you're in `wtypes.h` it is `long`, oops). That's walking a line that, on the surface, looks thin and it seems feasible that the return type of a lambda gets deduced incorrectly. An explicit return type would fix that (until you're getting the definition from `wtypes.h`), but who does that.

In practice... most `HRESULT` constants are defined so that their type is correctly casted as `long`/`HRESULT`. Additionally, `TRUE`/`FALSE` seem to be defined as just `1` and `0` respectively, meaning they will be `int`s and not `long`, regardless of how `BOOL` is defined. Additionally, if the user uses `true`/`false`, it'll take the "fallback" `else` below and will cast to `BOOL` correctly.

All this is to say, I think this is _probably_ okay, but it does give me pause. I can't think of any non-fabricated scenario that would cause issues (someone using `return 0` instead of `return S_OK` isn't a realistic fear in my eyes). Still, calling it out in case I'm overlooking something that someone else sees.